Add Futures to ProcessGroupGloo

agolynski opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57818 Add Futures to ProcessGroupGloo**

Differential Revision: [D28304171](https://our.internmc.facebook.com/intern/diff/D28304171)

agolynski committed:
Add Futures to ProcessGroupGloo

[ghstack-poisoned]

agolynski committed:
Update on "Add Futures to ProcessGroupGloo"




[ghstack-poisoned]

agolynski committed:
Update on "Add Futures to ProcessGroupGloo"




[ghstack-poisoned]

agolynski committed:
Update on "Add Futures to ProcessGroupGloo"




[ghstack-poisoned]

agolynski committed:
Update on "Add Futures to ProcessGroupGloo"




[ghstack-poisoned]

agolynski committed:
Update on "Add Futures to ProcessGroupGloo"




[ghstack-poisoned]

agolynski committed:
Update on "Add Futures to ProcessGroupGloo"




[ghstack-poisoned]

agolynski committed:
Update on "Add Futures to ProcessGroupGloo"




[ghstack-poisoned]

agolynski committed:
Update on "Add Futures to ProcessGroupGloo"




[ghstack-poisoned]

agolynski commented:
@agolynski has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28304171).

SciPioneer commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -420,13 +420,17 @@ const auto kLoopbackAddress = "127.0.0.1";
+
+  // FIXME: We need to call it here since Future completion requires all
+  // the work to be synchronized to CUDA.
+  work->synchronize();
```
Is this only specific to GLOO?

SciPioneer commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -435,11 +439,61 @@ std::vector<at::Tensor> ProcessGroupGloo::AsyncWork::result() {
+}
+
+namespace {
+c10::intrusive_ptr<c10::ivalue::Future> CreateFutureForOutput(
```
nit: AsOutput?

mrshenli replied:
ReturnFutureWithOutput -> returnFutureWithOutput to stay consistent with other function names in this file.

SciPioneer commented on torch/lib/c10d/test/ProcessGroupGlooTest.cpp:
```diff
@@ -204,6 +204,28 @@ std::vector<std::vector<at::Tensor>> waitWork(
+    try {
+      fut->wait();
+    } catch (const std::exception& ex) {
+      std::cerr << "Exception received: " << ex.what() << std::endl;
```
Shall we abort here?

wayi1 reviewed (approved):
Thanks for the merge!

mrshenli commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -435,11 +439,61 @@ std::vector<at::Tensor> ProcessGroupGloo::AsyncWork::result() {
+      c10::ListType::create(c10::TensorType::get()));
+}
+
+void ReturnFutureWithOutput(
```
ditto

mrshenli commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -435,11 +439,61 @@ std::vector<at::Tensor> ProcessGroupGloo::AsyncWork::result() {
+    std::vector<std::vector<at::Tensor>> outputTensors,
+    const char* profilingTitle,
+    const c10::optional<std::vector<at::Tensor>>& inputTensors)
+    : ProcessGroup::Work(-1, OpType::UNKNOWN, profilingTitle, inputTensors),
```
(This is prior to this PR): Curious, any reason for using UNKNOWN OpType here? I saw UNKNOWN is also used for recv, but I don't recall the reason for that.

agolynski replied:
I've looked around and seems like optype is only used in NCCL process group for p2p comm at #45873. Maybe there are other uses which I haven't found?

https://github.com/pytorch/pytorch/search?q=retrieveOpType

https://github.com/pytorch/pytorch/search?q=opType_

mrshenli commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -420,13 +420,17 @@ const auto kLoopbackAddress = "127.0.0.1";
-  work->finish(eptr);
+
+  // FIXME: We need to call it here since Future completion requires all
+  // the work to be synchronized to CUDA.
```
Since we now have CUDA support in `ivalue::Future`, should this synchronization work be done by `ivalue::Future` instead?  cc @lw

agolynski replied:
completely agree. 

BTW, we used to do some other things in synchronize() (e.g. copying tensors), however seems like now all derived versions of synchronize() match ivalue::Future::synchronizeWithCurrentStreams()

The plan is to deprecate work API in 1.10, so we don't have any more versions of synchronize() and then use
ivalue::Future::synchronizeWithCurrentStreams() directly.

cc @lw

mrshenli commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -435,11 +439,61 @@ std::vector<at::Tensor> ProcessGroupGloo::AsyncWork::result() {
+
+void ProcessGroupGloo::AsyncWork::finishWorkGlooError(std::exception_ptr eptr) {
+  future_->setError(eptr);
+  finish(eptr);
```
similarly, since `ivalue::Future` has similar concepts and states for `completed_` / `exception_`, shall we just use that instead? cc @lw

agolynski replied:
we exposed completed_  and exception_ as protected members of ProcessGroup::Work, so potentially they can be referred to by 3rd parties. Let's make deprecation announcement for 1.10 and then it'd be easier to make better Work/Future merge which can also include CUDA synchronization.

lw replied:
I don't get it: how can third-parties access the `completed_` and `exception_` fields? If they are protected it means they can only be accessed by subclasses, but I don't see the point of users subclassing Work: even if they did, ProcessGroupGloo would still return its own classes, and not the user ones, so what does one gain? Moreover, the trailing underscore in the fields names is C++'s "convention" for private fields. All in all, I don't think we should worry about users depending on it.

I agree with Shen that now, effectively, we're doing the same thing twice within Work: we have two ways to mark completion (Work's own one, and the one within Future), etc. One way to go about it is to redefine Work so that it just "forwards" each of its methods to the Future it holds (e.g., Work::wait() basically just calls this->future_->wait()).

agolynski replied:
@lw 

> I don't get it: how can third-parties access the `completed_` and `exception_` fields? If they are protected it means they can only be accessed by subclasses, but I don't see the point of users subclassing Work: even if they did, ProcessGroupGloo would still return its own classes, and not the user ones, so what does one gain?

MPI/GLOO subclass Work, same way 3rd party implementers can also subclass Work and use `completed_` and `exception_`?

ProcessGroupGloo ProcessGroupMPI Work subclasses use (and 3rd party can) PG::Work's `completed_` and `exception_`, so if we remove them 3rd party Work classes will break.

> Moreover, the trailing underscore in the fields names is C++'s "convention" for private fields. 

I thought trailing underscore just means it's a member, not necessarily private?

> All in all, I don't think we should worry about users depending on it.

I think you are right, but it's good to be cautious.

> I agree with Shen that now, effectively, we're doing the same thing twice within Work: we have two ways to mark completion (Work's own one, and the one within Future), etc. One way to go about it is to redefine Work so that it just "forwards" each of its methods to the Future it holds (e.g., Work::wait() basically just calls this->future_->wait()).

the plan is to remove fields from PG::Work and replace it a future. Note that right now, MPI and GLOO subclasses hold a future.

lw replied:
Right, I hadn't understood the `completed_` and `exception_` fields came from the root "abstract" base class Work (which, well, isn't actually abstract it seems). Indeed, people who created their own ProcessGroup implementations (as extensions) would be broken by any change there. But I think we eventually will need to do it, right?

mrshenli commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -435,11 +439,61 @@ std::vector<at::Tensor> ProcessGroupGloo::AsyncWork::result() {
+
+void ProcessGroupGloo::AsyncWork::finishWorkGloo() {
+  ReturnFutureWithOutput(future_, outputTensors_);
+  finish();
```
ditto

agolynski replied:
let's leave it for 1.10?

mrshenli reviewed (commented):
Thanks for adding this! I left some questions for consolidating Work/Future APIs/fields. I am OK with addressing some of those comments in follow up PRs, but let's not stop here can call the consolidation done.  

One example is that, after the consolidation, at least we shouldn't have both Work and Future keeping `completed_` / `exception_` field. 

cc @lw

agolynski committed:
Update on "Add Futures to ProcessGroupGloo"


Differential Revision: [D28304171](https://our.internmc.facebook.com/intern/diff/D28304171)

[ghstack-poisoned]

agolynski committed:
Update on "Add Futures to ProcessGroupGloo"


Differential Revision: [D28304171](https://our.internmc.facebook.com/intern/diff/D28304171)

[ghstack-poisoned]

lw commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -435,11 +439,61 @@ std::vector<at::Tensor> ProcessGroupGloo::AsyncWork::result() {
+}
+
+void returnFutureWithOutput(
+    c10::intrusive_ptr<c10::ivalue::Future> future,
```
Nit: no need to pass the pointer by value, this function could just take a `Future&` and thus avoid the copy and the consequent refcount increase+decrease.

lw commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -435,11 +439,61 @@ std::vector<at::Tensor> ProcessGroupGloo::AsyncWork::result() {
+namespace {
+c10::intrusive_ptr<c10::ivalue::Future> createFutureAsOutput(
+    const std::vector<std::vector<at::Tensor>>& outputTensors) {
+  if (outputTensors.size() > 1) {
+    return c10::make_intrusive<c10::ivalue::Future>(
+        c10::ListType::create(c10::ListType::create(c10::TensorType::get())));
+  }
+  return c10::make_intrusive<c10::ivalue::Future>(
+      c10::ListType::create(c10::TensorType::get()));
```
Is there a reason for this? I find it odd that a type information depends on a runtime condition. Types are typically useful to "predict what would happen before it happens" (e.g., static analysis) but by doing this we defeat that purpose.

lw replied:
I'm not very knowledgeable on the PG API. My first impression is that it would be more "regular" and predictable if here we always used a list of tensors, even in case of a singleton.

If however you say that different methods of a PG need to return different types, that's fine, and some can return a list of tensors while other can return a singleton tensor. However I'd like if a certain method always returns the same type, no matter its arguments.

agolynski replied:
agreed. I don't mind returning a singleton tensor wrapped in a list and we already do, but how would we deal with returning list of lists?

lw reviewed (commented):
It's a good first step to add the `getFuture` method to Gloo too, but we agree that the final solution will be to remove that method and to make Work itself be a Future subclass right?

agolynski committed:
Update on "Add Futures to ProcessGroupGloo"


Differential Revision: [D28304171](https://our.internmc.facebook.com/intern/diff/D28304171)

[ghstack-poisoned]

rohan-varma reviewed (commented):
A bit late to this, but are we planning to also add support for send/recv futures and testing? It looks like we would need to add that to Send/RecvWork.

agolynski commented:
> A bit late to this, but are we planning to also add support for send/recv futures and testing? It looks like we would need to add that to Send/RecvWork.

yes, we could add handling that 1.10. Send/RecvWork (both GLOO and MPI) are running in async mode, so they would require some more work (maybe sacrificing some efficiency or API or a combination)
